### ISSUE TITLE
Validation binding dependencies

### ISSUE THREAD

**Issue by heruan** [aurelia/validation#279]
The old library had the ability to make rules dependent on others, i.e. re-evaluate a binding when other bindings are updated; now, this won't work:
```html
<input type="password" value.bind="password & validate">
<input type="password" value.bind="passwordConfirm & validate">
```
assume `passwordConfirm` must be equal to `password` (supported for example in `aurelia-validatejs` with `@equality`) the `passwordConfirm & validate` binding should trigger also when `password` changes.

**Comment by heruan** [comment]
Some ideas:
- the `validate` binding behavior currently accept a `rules` argument, which is very useful or even necessary in certain cases; so, I would not take the road of another argument to pass to keep a clean signature;
- I would instead create a `ValidationRule` interface, to have a `dependsOn` array of other `ValidationRule`s;
- then, the binding on the dependent property would subscribe its update/validation also on its rules dependencies.

**Comment by PWKad** [comment]
This was previously supported because all rules of the object were re-evaluated after any property of that object changed.  Perhaps if the code for `equality` is similar to how it was previously implemented we could use that to trigger a dependency on the target property under the covers.

**Comment by heruan** [comment]
I came up with https://github.com/aurelia/validation/pull/280 and https://github.com/aurelia/validatejs/pull/117. I'll publish a GistRun to see it in action!

**Comment by heruan** [comment]
Of course, all that fuss could be avoided if we call `validate()` to re-evaluate all rules when a binding triggers, instead of evaluate just the rule for its property. Too much overhead?

**Comment by jdanyow** [comment]
I'm hesitant to put UI stuff in the fluent rule api because it's intended to be used server-side. Would it be better to create some sort of binding behavior that you can use to express UI triggers for rule validation?

```html
<input type="password" value.bind="password & validate">
<input type="password" value.bind="passwordConfirm & validate & dependsOn:'password'">
```

Today you could do something like this:

```html
<div change.delegate="controller.validate()">
  <input type="password" value.bind="password & validate">
  <input type="password" value.bind="passwordConfirm & validate">
</div>
```

Or this:

```html
<input type="password" value.bind="password & validate"
       change.delegate="controller.validate()">
<input type="password" value.bind="passwordConfirm & validate">
```

Or this:

```html
<input type="password" value.bind="password & validate"
       change.delegate="controller.validate({ object: $this, propertyName: 'passwordConfirm' })">
<input type="password" value.bind="passwordConfirm & validate">
```

**Comment by dkent600** [comment]
It seems to me that any rule that depends on another rule should be re-defined using `.satisfies()`, no?  I'm not clear what is the use case where this would not be appropriate or possible.

But using `satisfies` still presents an issue.

My use case:  

I am validating at the object level, thus must manually call `ValidationController.validate()`.  

Each object contains object rules (defined by `ensureObject().satisfies()`).  These object rules do not depend on other rules, but they do relate to a set of properties on the object.  However, unlike with `ensure().satisfies()` it is not possible to associate `ensureObject().satisfies()` with a property name, much less with multiple property names.  

Further, I am using neither the `validate` binding nor a `ValidationRenderer`.  Again, I am manually calling `ValidationController.validate()`, and manually rendering the error information when `validate` completes.

A ValidationRenderer could be useful for me in this scenario, but it is not, for the following reasons:

1) without using the `validate` binding, it does not provide the element(s) with which a property is associated.
2) it provides neither property names nor elements for rules define by `ensureObject().satifies()` 

Thus I propose the following:

Add a parameter `propertyNames` to `ensureObject().satisfies(instance)`, like this:  `ensureObject().satisfies(instance, propertyNames?: string[])` 

Change the type of `ValidationError.propertyName` from `string` to `string | Array<string>` (or have it always be an array). 
Perhaps change the name of `ValidationError.propertyName` to the plural `ValidationError.propertyNames`

I think this fits fairly nicely into the existing paradigm.

See will help with this (feature request)[https://github.com/aurelia/validation/issues/351]

**Comment by dkent600** [comment]
Hmm, I just realized there is another aspect to this:

I noted above "1) without using the validate binding, [ValidationRenderer] does not provide the element(s) with which a property is associated."

Adding property names to satisfied does not solve this problem.

I don't want to have to add a `validate` binding because I am *always* calling `ValidationController.validate()` and don't want to validate twice.

How to address this bit?  Gotta think about that...

**Comment by dkent600** [comment]
Alternative suggestion:

Enable `ensure(a => a.propertyname)` (or `ensure(propertyname`) to take multiple property names instead of just one.

**Comment by rajajhansi** [comment]
@jdanyow, I tried to use aurelia-validation on the server side in a nodejs application but it looks like ValidationRules has still dependencies on Element (lib.d.ts aka DOM dependency), which makes it fail. I noticed that the documentation page on "Server Side Validation"(http://aurelia.io/hub.html#/doc/article/aurelia/validation/latest/validation-basics/13)  isn't ready yet because I guess it is still work in progress. I've had a great success using aurealia-dependency-injection on the server side without any issues. I'm about to wrap up part 7 of my tutorial in my blog ( http://opensourcedotnetter.blogspot.com/) on Aurelia validation. Please advice if I should wait for a release of aurelia-validation that works on the server side or should I go ahead with using validate-js directly or the deprecated aurelia-validatejs package. Thanks.

**Comment by rajajhansi** [comment]
I face the same issue using aurelia-validatejs package (although it is deprecated) as well since it has a dependency on aurelia-validation. I'm going to go ahead and use validatejs directly in nodejs and return the validation errors to the client.

**Comment by jdanyow** [comment]
Comment from @dkent600 here: https://github.com/aurelia/validation/issues/351#issuecomment-269735880 has gist with a couple of scenarios.

**Comment by dkent600** [comment]
I note that in addition to adding the property names to `satisfies`, we should probably also add them to `customRule`, though I haven't confirmed with testing.

**Comment by dkent600** [comment]
@heruan I'm looking into this, particular at the original use case you provided, and am wondering, is it sufficient when you set up your rules to simply require that both `password` and `confirmPassword` satisfy a custom or tagged rule that you will have created to match the two values?  If not, what more is needed?  Thanks.

**Comment by dkent600** [comment]
See also this comment, with a gist:  https://github.com/aurelia/validation/issues/351#issuecomment-275755848
